WordPress/gutenberg
# Site Editor: Add new 'Push changes to Global Styles' button

## What?
Alternative to https://github.com/WordPress/gutenberg/pull/45961. Closes https://github.com/WordPress/gutenberg/issues/44361.

Adds a new 'Push changes to Global Styles' button to the Advanced panel for all blocks in the Site Editor. Pressing it applies the block's typography, spacing, dimensions, and color styling to all blocks of that type in Global Styles.

## Why?
See https://github.com/WordPress/gutenberg/issues/44361.

## How?
A new `editor.BlockEdit` filter is added to `@wordpress/edit-site`. This filter renders the button in an `AdvancedControls` fill.

When the button is pressed, we iterate through the styles that are supported by the selected block type in Global Styles, and, for each one: 1) use `setAttribute()` to unset the block style; 2) use `GlobalStylesContext` to move the style to Global Styles for that block type.

It's difficult to do this in one undo level using the current `@wordpress/core-data` API so, for now, we create no undo levels and implement a custom Undo button in a snackbar notification.

## Testing Instructions
A basic E2E test is included. To test manually:

1. Open the site editor by navigating to Appearance → Editor.
2. Select a block e.g. a Heading block.
3. Make some stylistic changes in the block inspector.
4. Open the _Advanced_ panel and click _Push changes to Global Styles_.
5. The typography, spacing, dimensions and color changes you made should now apply to all blocks of that type ( e.g.Heading). You can verify this by navigating to Global Styles → Blocks → e.g. Heading.

## Follow up work

- Add the button to the post editor. See https://github.com/WordPress/gutenberg/pull/46446#issuecomment-1352536992
- Use `@wordpress/core-data` undo/redo functionality instead of rolling our own

---

_Note: @noisysocks wrote this description, so don't judge @ramonjd if there are typos 😀_
